onnxruntime/contrib_ops/cuda/bert/attention_impl.cu

```diff
@@ -119,6 +119,21 @@ __device__ inline void SoftmaxSmall(const int sequence_length, const int valid_l
 
   const int num_valid = is_unidirectional ? (blockIdx.x % sequence_length) + 1 : valid_length;
 
+  // If `num_valid` is 0, it means that mask for this sample in the batch is 0 
+  // (valid_length is 0 in the step above)
+  // If we reach this step, sequence_length is guaranteed to be > 0 
+  // (we ensure this in the beginning of the op computation).
+  // Hence, we stop at this stage to prevent unnecessary steps below given that there are no 
+  // values to compute softmax over. We assign the output an arbitrary value of 0
+  // which is the best we can do considering that it is hard to perform attention by considering
+  // 0 entities in the given sequence
+  // This guard is primarily being added to avoid bad out-of-the-box experience like seeing NaNs
+  // like in Github issue #3977
+  if (num_valid == 0) {
```

Comment:
No need to add a similar check in Softmax() as it is implicitly done with this ternary check - 

`const float val = (i < num_valid) ? expf(float(input[index]) - max_block) * sum_reverse_block : 0.f;`

If num_valid is 0, we won't ever evaluate i < num_valid to be true and hence won't see NaNs and assign 0 to the output